src/client/client.rs

```diff
@@ -156,107 +69,24 @@ where
     /// If the connection to the broker fails, method returns Err variable that contains
     /// Reason codes returned from the broker.
     pub async fn connect_to_broker<'b>(&'b mut self) -> Result<(), ReasonCode> {
```

Comment:
Honestly, I am not sure what I prefer. Out of curiosity, I checked

* *MQTTNet* (a C# library I use) and it [uses explicit connect](https://github.com/dotnet/MQTTnet/blob/f5d9b65139ef18cc3deef26f41c9e6a994c4324a/Samples/Client/Client_Connection_Samples.cs#L38)
* *rumqtt* and it [uses connect on construct](https://github.com/bytebeamio/rumqtt/blob/f460805c823462a6b9f4460a6cdf1bac76094814/rumqttc/examples/asyncpubsub_v5.rs#L15)

so that did not help.

"Connect on construct" would be perfect if it can guarantee that you are then always connected, but it can happen that the broker disconnects with you and then you can't publish messages either. One thing that may be an advantage of splitting construction and connection is that the client can be initialized without using async, which may be useful in some contexts. Another advantage could be that you can disconnect and connect again later with the same instance of client.

A compromise could be providing both, constructor without connecting *and* connecting constructor. The constructor without connect could be named such that it is obvious that it does not connect.